### PR TITLE
DNS server support

### PR DESCRIPTION
**Pull request by timboudreau**
Support for writing DNS servers, and various API changes to accomplish that, specifically:

 - `DnsType` -> `DnsType<RecordType>` where RecordType is a subclass of `DnsEntry`, and has payload decoders for specific types
 - Introduced record-specific subclasses of `DnsEntry` for common types, with fallback to the ByteBuf-based `DnsResource`
 - `DnsEntry` gains public `writeTo()` and `writePayload()` to write itself correctly into a ByteBuf
 - Added `DnsQueryDecoder` and `DnsResponseEncoder`
 - Introduced `NameWriter` for writing names, so that it's possible to implement a stateful `NameWriter` to do DNS name compression
 - `DnsMessage` is threadsafe (needed for proxy DNS servers where multiple calls to remote DNS servers may contribute partial responses)
    - If someone really feels strongly that creating the `List` instances it uses internally should be lazy, that could be done, but it would be good to have some evidence that really accomplishes something - feels like premature optimization
 - `DnsQuestion` is no longer a subclass of `DnsEntry` (which now has a time-to-live, since that is common to all other record types).
 - Added the package io.netty.handler.codec.dns.server with `DnsServerHandler`, a `SimpleChannelInboundHandler<DatagramPacket>` that handles the common plumbing of writing a DNS server that is unlikely to differ between implementations
 - Corrected error codes in `DnsResponseCode` to match https://tools.ietf.org/html/rfc2929 (they were sequential, but in fact 11-15 are unused and numbering restarts at 16)
 - Added error codes that exist but were not included in `DnsResponseCode` but are part of the spec
 - Added `DnsDecoderException` which can specify a `DnsResponseCode`
 - `DnsQuery` implements `Iterable<DnsQuestion>`, simply because it's intuitive
 - Reasonable `toString()` implementations on entity classes for logging purposes
 - Added to netty-examples:  DNS server, DNS client, proxy DNS server

### CODE REVIEW

**Comment by nezhazheng** [comment]
any progress?

**Comment by trustin** [comment]
@nezhazheng Sorry, but no. On my radar though.

File: codec-dns/src/main/java/io/netty/handler/codec/dns/Ipv4AddressRecord.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            sb.append(ints[i]);
+            if (i != ints.length - 1) {
+                sb.append('.');
+            }
+        }
+        return sb.toString();
+    }
+
+    private static int[] toInts(int val) {
+        int[] ret = new int[4];
```
Comment (bk1te) [inline codec-dns/src/main/java/io/netty/handler/codec/dns/Ipv4AddressRecord.java]: NetUtil.intToIpAddress()

**Comment by timboudreau** [comment]
So, this got integrated, sort of?

**Comment by normanmaurer** [comment]
@timboudreau no I think this was closed because we renamed master to master-deprecated branch :( Could you reopen against master-deprecated or even better 4.1 ?

**Comment by timboudreau** [comment]
I don't see any obvious way to reopen it (don't know if I don't have permissions for that or if GitHub's UI for it is non-obvious).  Or you want me to file a new issue w/ new branch and new patches?

**Comment by trustin** [comment]
I didn't have time for taking care of this PR but I do hope to. Let me take a stab this weekend.

**Comment by timboudreau** [comment]
Thanks!

On Thu, Dec 17, 2015 at 10:57 AM, Trustin Lee <notifications@github.com>
wrote:

> I didn't have time for taking care of this PR but I do hope to. Let me
> take a stab this weekend.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/netty/netty/pull/3146#issuecomment-165492808>.
>



-- 
http://timboudreau.com

**Pull request closed without merging**